fence/man/fence_ifmib.8
.TH fence_ifmib 8

.SH NAME
fence_ifmib - I/O Fencing agent for IF-MIB capable SNMP devices

.SH SYNOPSIS
.B
fence_ifmib
[\fIOPTION\fR]...

.SH DESCRIPTION
fence_ifmib is an I/O Fencing agent which can be used with any SNMP IF-MIB capable
device. Agent internally uses snmpget, snmpset and snmpwalk command.

It was written with managed ethernet switches in mind, in order
to fence iSCSI SAN connections.  However, there are many devices that support
the IF-MIB interface.  The agent uses IF-MIB::ifAdminStatus to control the
state of an interface.

fence_ifmib accepts options on the command line as well as from stdin.
Fenced sends parameters through stdin when it execs the agent.  fence_ifmib
can be run by itself with command line options.  This is useful for testing.

.SH OPTIONS
.TP
\fB-a\fP \fIIPaddress\fR
IP address or hostname of the SNMP device. Can be used any syntax supported by snmpget.
.TP
\fB-h\fP
Print out a help message describing available options, then exit.
.TP
\fB-c\fP \fIcommunity\fR
The read/write community string to be used in the request.
.TP
\fB-n\fP \fIname\fR
Name of port to fence (fc1/1) or ifIndex (43).
.TP
\fB-p\fP \fIpassword\fR
Password for login for SNMP v3 (authentication protocol pass phrase).
.TP
\fB-P\fP \fIpassword\fR
Password for privacy for SNMP v3 (privacy protocol password).
.TP
\fB-S\fP \fIscript\fR
Script to run to retrieve password for login for SNMP v3 (authentication protocol pass phrase).
.TP
\fB-R\fP \fIscript\fR
Script to run to retrieve privacy for SNMP v3 (privacy protocol password).
.TP
\fB-l\fP \fIlogin\fR
Login name for SNMP v3 (security name).
.TP
\fB-d\fP \fIversion\fR
SNMP version (1,2c,3).
.TP
\fB-b\fP \fIauth_protocol\fR
SNMP authentication protocol (MD5|SHA).
.TP
\fB-E\fP \fIsec_level\fR
SNMP security level (noAuthNoPriv|authNoPriv|authPriv).
.TP
\fB-B\fP \fIpriv_protocol\fR
SNMP privacy protocol (DES|AES).
.TP
\fB-u\fP \fIudp_port\fR
UDP/TCP port to use.
.TP
\fB-o\fP \fIaction\fR
The action required.  off (default), on, status, list or monitor. Deprecated
options (enable -> on and disable -> off) can be used too.
.TP
\fB-v\fP
Verbose. Record session to stdout, or debug file if specified (see -D).
.TP
\fB-D\fP
Specifies file, where will be written debug messages from session.
.TP
\fB-V\fP
Print out a version message, then exit.

.SH STDIN PARAMETERS
.TP
\fIagent = < param >\fR
This option is used by fence_node(8) and is ignored by fence_ifmib.
.TP
\fIipaddr = < param >\fR
IP address or hostname of the SNMP device. Can be used any syntax supported by snmpget.
.TP
\fIcommunity = < param >\fR
The read/write community string to be used in the request.
.TP
\fIport = < param >\fR
Name of port to fence (fc1/1) or ifIndex (43).
.TP
\fIpasswd = < param >\fR
Password for login for SNMP v3 (authentication protocol pass phrase).
.TP
\fIsnmp_priv_passwd\fR
Password for privacy for SNMP v3 (privacy protocol password).
.TP
\fIpasswd_script = < param >\fR
Script to run to retrieve password for login for SNMP v3 (authentication protocol pass phrase).
.TP
\fIsnmp_priv_passwd_script = < param>\fR
Password for privacy for SNMP v3 (privacy protocol password).
.TP
\fIlogin = < param >\fR
Login name for SNMP v3 (security name).
.TP
\fIsnmp_version = < param >\fR
SNMP version (1,2c,3).
.TP
\fIsnmp_auth_prot = < param >\fR
SNMP authentication protocol (MD5|SHA).
.TP
\fIsnmp_sec_level = < param >\fR
SNMP security level (noAuthNoPriv|authNoPriv|authPriv).
.TP
\fIsnmp_priv_prot = < param >\fR
SNMP privacy protocol (DES|AES).
.TP
\fIudpport = < param >\fR
UDP/TCP port to use.
.TP
\fIaction = < param >\fR
The action required.  off (default), on, status, list or monitor. Deprecated
options (enable -> on and disable -> off) can be used too.
.TP
\fIverbose = < param >\fR
Verbose.  Record session to stdout, or debug file if specified (see debug).
.TP
\fIdebug = < param >\fR
Specifies file, where will be written debug messages from session.

.SH SEE ALSO
fence(8), fence_node(8)


fence/agents/ifmib/fence_ifmib.py
#!/usr/bin/python

# The Following agent has been tested on:
# - Cisco MDS UROS 9134 FC (1 Slot) Chassis ("1/2/4 10 Gbps FC/Supervisor-2") Motorola, e500v2
#   with BIOS 1.0.16, kickstart 4.1(1c), system 4.1(1c)
# - Cisco MDS 9124 (1 Slot) Chassis ("1/2/4 Gbps FC/Supervisor-2") Motorola, e500
#   with BIOS 1.0.16, kickstart 4.1(1c), system 4.1(1c)
# - Partially with APC PDU (Network Management Card AOS v2.7.0, Rack PDU APP v2.7.3)
#   Only lance if is visible

import sys, re, pexpect
from fencing import *
from fencing_snmp import *

#BEGIN_VERSION_GENERATION
RELEASE_VERSION="IF:MIB SNMP fence agent"
REDHAT_COPYRIGHT=""
BUILD_DATE=""
#END_VERSION_GENERATION

### CONSTANTS ###
# IF-MIB trees for alias, status and port
ALIASES_OID=".1.3.6.1.2.1.31.1.1.1.18"
PORTS_OID=".1.3.6.1.2.1.2.2.1.2"
STATUSES_OID=".1.3.6.1.2.1.2.2.1.7"

# Status constants returned as value from SNMP
STATUS_UP=1
STATUS_DOWN=2
STATUS_TESTING=3

### GLOBAL VARIABLES ###
# Port number converted from port name or index
port_num=None

### FUNCTIONS ###

# Convert port index or name to port index
def port2index(conn,port):
	res=None

	if (port.isdigit()):
		res=int(port)
	else:
		ports=conn.walk(PORTS_OID,30)

		for x in ports:
			if (x[1].strip('"')==port):
				res=int(x[0].split('.')[-1])
				break

	if (res==None):
		fail_usage("Can't find port with name %s!"%(port))

	return res

def get_power_status(conn,options):
	global port_num

	if (port_num==None):
		port_num=port2index(conn,options["-n"])

	(oid,status)=conn.get("%s.%d"%(STATUSES_OID,port_num))
	return (status==str(STATUS_UP) and "on" or "off")

def set_power_status(conn, options):
	global port_num

	if (port_num==None):
		port_num=port2index(conn,options["-n"])

	conn.set("%s.%d"%(STATUSES_OID,port_num),(options["-o"]=="on" and STATUS_UP or STATUS_DOWN))

# Convert array of format [[key1, value1], [key2, value2], ... [keyN, valueN]] to dict, where key is
# in format a.b.c.d...z and returned dict has key only z
def array_to_dict(ar):
	return dict(map(lambda y:[y[0].split('.')[-1],y[1]],ar))

def get_outlets_status(conn, options):
	result={}

	res_fc=conn.walk(PORTS_OID,30)
	res_aliases=array_to_dict(conn.walk(ALIASES_OID,30))

	for x in res_fc:
		port_num=x[0].split('.')[-1]

		port_name=x[1].strip('"')
		port_alias=(res_aliases.has_key(port_num) and res_aliases[port_num].strip('"') or "")
		port_status=""
		result[port_name]=(port_alias,port_status)

	return result

# Define new options
def ifmib_define_defaults():
	all_opt["snmp_version"]["default"]="2c"

# Main agent method
def main():
	global port_oid

	device_opt = [ "help", "version", "agent", "quiet", "verbose", "debug",
		       "action", "ipaddr", "login", "passwd", "passwd_script",
		       "test", "port", "separator", "no_login", "no_password",
		       "snmp_version", "community", "snmp_auth_prot", "snmp_sec_level",
		       "snmp_priv_prot", "snmp_priv_passwd", "snmp_priv_passwd_script",
		       "udpport"]

	atexit.register(atexit_handler)

	options=process_input(device_opt)

	# Emulate enable/disable functionality
	if (options.has_key("-o")):
		options["-o"]=options["-o"].lower()

		if (options["-o"]=="enable"):
			options["-o"]="on"
		if (options["-o"]=="disable"):
			options["-o"]="off"
	else:
		options["-o"]="off"

	options = check_input(device_opt, options)

	# Operate the fencing device
	fence_action(FencingSnmp(options), options, set_power_status, get_power_status, get_outlets_status)

if __name__ == "__main__":
	main()
